### PR TITLE
Further remove code duplication between SpectrumDataset and MapDataset

### PR DESCRIPTION
**Pull request by adonath**
**Description**

This pull request further remove code duplication between `SpectrumDataset`, `SpectrumDatasetOnOff`, `MapDataset` and `MapDatasetOnOff`. It implements the following changes:

- Make `SpectrumDatasetOnOff` inherit from `MapDatasetOnOff`
- Move common plotting functionality between the spectrum dataset to a common "mixin" class
- Unify `.create()` methods between `MapDataset` and `SpectrumDataset`
- Move `.energy_range` to `MapDataset` and make it work for cubes as well
- Replace `SpectrumDataset.plot_residuals()` by `.plot_residuals_spectral()` (which is now inherited from `MapDataset`)

The only remaining functionality on the spectrum datasets is now plotting and serialisation. I would like to do one follow up PR to move the serialisation to `datasets/io.py` and implement a minimal registry pattern.

If we manage to further unify the plotting between the `SpectrumDataset` and `MapDataset`, the `SpectrumDataset` and `SpectrumDatasetOnOff` could be removed completely along with the `SpectrumDatasetMaker`. This would finally remove another few hundred lines of code without removing any functionality.

ToDo:
- [x] Adapt tutorials
- [x] Fix remaining test fails

### CODE REVIEW

**Comment by adonath** [comment]
Here is a list of methods, that are useless or poorly defined on a `MapDataset` with a `RegionGeom`:

- `MapDataset.cutout()`, doesn't make sense if there is just one spatial bin. I think it can never be defined in a meaningful way. So it should just raise an error.
- `MapDataset.plot_residuals_spatial()`. Again this will never be defined, however one could adapt the visualisation of `.plot_residuals()` to the geometry of the contained data instead, or just raise and error...
- `MapDataset.to_spectrum_dataset()` would just return `self`. This is not elegant, but I think it's well defined and acceptable behaviour.
- `MapDataset.pad()` is not well defined, however it would be useful for a `RegionGeom` if it had a `axis_name` argument
- `MapDataset.to_image()` is bit of a misnomer,  however the behaviour (integration over the energy axis) is well defined and useful for a `RegionGeom` as well.

I think the behaviour of all the other methods on the `MapDataset` is well defined, even if not fully implemented. We mostly miss a serialisation format for region geometries with extra axes, so that things like `.from_hdulist()`, `.to_hdulist()`, `.read()` and `.write()` etc. work.  Looking at the list of methods above I'm not 100% whether it justifies keeping the `SpectrumDataset` and `SpectrumDatasetOnOff` as a separate abstraction...

I realise handling the 1D analysis as a special case of the 3D one might cause confusion, however already now the API is 90% the same and to me the advantages are more important:

- Only a single API to learn
- Only one implementation to maintain
- Containment correction for PSF and morphology  "on the fly"

In practice the difference rarely appears in the user facing API. Beside the cases I mentioned above, the data is stored in a 3D Numpy array, and users need to access the 1D array like `.data[:, 0, 0]`. Or `.data.squeeze()`. But this is the case already now...

**Comment by adonath** [comment]
Conclusion from the meeting today was:
- First remove the code duplication, but keep the "safe" API. I.e. move common functionality to a common hidden base class and just implement the functionality that is well defined...
- In the meantime @LauraOlivera and @fabiopintore are implementing the support to integrate a model or effective area using a `RegionGeom`

Further cleanup and unification can then follow step by step...

**Comment by adonath** [comment]
I realised the proposal with common base class is not a straightforward as it seems. We need to introduce both, something like `MapDatasetBase` and a `MapDatasetOnOffBase`. This means we finally end up with 6! classes plus a `PlotMixin` for the spectrum dataset classes. From a maintainer / contributor perspective I think this is hard to justify compared to 2 classes if we go for a full unification...I'll still go ahead and implement this, but in general I'm not yet happy with this...

**Comment by adonath** [comment]
As stated in  my previous comment the solution with the common base classes would introduce a complex class hierarchy with multiple inheritance etc. I decided to stick with the previous implementation and have the simpler model of `SpectrumDataset` inheriting from `MapDataset` and `SpectrumDatasetOnOff` inheriting from `SpectrumDatasetOnOff`, the common plotting of the spectral dataset is in the `PlotMixin` class, which is not visible to users...

**Review by registerrier — APPROVED** [review]
Thanks @adonath !

This looks good. Of course it is a bit complex to review given the length. 
Only comments are inline: clarify read/write with an error message to use `to_ogip` methods.
And the default `mask_safe` in create.

Generally, the only API change is the `SpectrumDataset.create` which is fine. Probably it would have been better to have it for 0.18, but it is a limited change. It mostly forces 1D users to learn the `RegionGeom` API, which is probably acceptable.

File: gammapy/datasets/tests/test_spectrum.py [hunk 1]
```diff
@@ -387,29 +392,38 @@ def setup(self):
...
         assert empty_dataset.energy_range[0] is None
 
     def test_create_stack(self):
-        stacked = SpectrumDatasetOnOff.create(self.e_reco, self.e_true)
+        geom = RegionGeom(region=None, axes=[self.e_reco])
+
+        stacked = SpectrumDatasetOnOff.create(
+            geom=geom, energy_axis_true=self.e_true
+        )
+        stacked.mask_safe.data += True
```
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/datasets/tests/test_spectrum.py:426]: This is what we do by default in `MapDataset` as well. I guess it's the more safe behaviour for stacking, which is the main use-case for `.create()` I would say...

File: gammapy/datasets/spectrum.py [hunk 2]
```diff
@@ -456,258 +200,41 @@ def peek(self, fig=None):
...
-        Parameters
-        ----------
-        npred_background : `~gammapy.maps.RegionNDMap`
-            Predicted background to be used in the on region.
-        random_state : {int, 'random-seed', 'global-rng', `~numpy.random.RandomState`}
-            Defines random number generator initialisation.
-            Passed to `~gammapy.utils.random.get_random_state`.
-        """
-        random_state = get_random_state(random_state)
+    def read(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/datasets/spectrum.py:210]: The `SpectrumDataset` currently does no support any I/O and in `SpectrumDatasetOnOff.read()` we already refer to `.to_ogip()`...

**Comment by adonath** [comment]
Thanks @registerrier, I can address any remaining comments in follow up PRs as well.

**Pull request merged**